mobilecoind/src/service.rs

```diff
@@ -581,12 +603,31 @@ impl<T: BlockchainConnection + UserTxConnection + 'static, FPR: FogPubkeyResolve
         &mut self,
         request: mc_mobilecoind_api::CreateTransferCodeRequest,
     ) -> Result<mc_mobilecoind_api::CreateTransferCodeResponse, RpcStatus> {
-        if request.root_entropy.len() != 32 {
+        // Only allow one type of entropy.
+        if !request.bip39_entropy.is_empty() && !request.root_entropy.is_empty() {
             return Err(RpcStatus::new(
                 RpcStatusCode::INVALID_ARGUMENT,
-                Some("root_entropy".to_string()),
+                Some("bip39_entropy/root_entropy".to_string()),
             ));
         }
+
+        // If we were provided with bip39 entropy, ensure it is 32 bytes long.
+        if !request.bip39_entropy.is_empty() && request.bip39_entropy.len() != 32 {
+            return Err(RpcStatus::new(
+                RpcStatusCode::INVALID_ARGUMENT,
+                Some("bip39_entropy".to_string()),
+            ));
+        }
+
+        // If we were provided with bip39 entropy, ensure it is 32 bytes long.
+        if !request.root_entropy.is_empty() && request.root_entropy.len() != 32 {
```

Comment:
What if both `root_entropy` and `bip39_entropy` are empty?

Reply:
Good catch.